File: Ser2023/Company/Company.Web/Modules/Store/WaresOrder/WaresOrderForm.cs
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace Company.Store.Forms;

[FormScript("Store.WaresMovement")]
[BasedOnRow(typeof(WaresOrderRow), CheckNames = true)]
public class WaresOrderForm
{
    [Tab("General")]
    [Category("Order")]
    [HalfWidth]
    public int? OperationTypeId { get; set; }
    [DefaultValue("now")]
    [HalfWidth]
    public DateTime OrderDate { get; set; }
    [HalfWidth]
    public String ShopId { get; set; }
    [HalfWidth]
    public String CounterpartyId{ get; set; }
    [HalfWidth]
    public Int32 EmployeeId { get; set; }
    [HalfWidth]
    public DateTime RequiredDate { get; set; }
    [HalfWidth]
    public Int32 ShipperId { get; set; }
    [HalfWidth]
    public DateTime ShippedDate { get; set; }
    [OneThirdWidth]
    public Int32 DocumentTypeId { get; set; }
    [OneThirdWidth]
    public String DocumentNumber { get; set; }
    [OneThirdWidth]
    public DateTime DocumentDate { get; set; }
    [OneThirdWidth]
    public Decimal Total { get; set; }

    [Category("WaresMovementDetailsRow")]
    [WaresOrderDetailEditor]
    public List<WaresOrderDetailRow> DetailList { get; set; }
}

File: Ser2023/Company/Company.Web/Modules/ServerTypes/Store/WaresOrderForm.ts
import { LookupEditor, DateEditor, StringEditor, DecimalEditor, Widget, PrefixedContext } from "@serenity-is/corelib";
import { CounterpartiesEditor } from "@/Store/Counterparties/CounterpartiesEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface WaresOrderForm {
    OperationTypeId: LookupEditor;
    OrderDate: DateEditor;
    ShopId: LookupEditor;
    CounterpartyId: CounterpartiesEditor;
    EmployeeId: LookupEditor;
    RequiredDate: DateEditor;
    ShipperId: LookupEditor;
    ShippedDate: DateEditor;
    DocumentTypeId: LookupEditor;
    DocumentNumber: StringEditor;
    DocumentDate: DateEditor;
    Total: DecimalEditor;
    DetailList: Widget;
}

export class WaresOrderForm extends PrefixedContext {
    static formKey = 'Store.WaresMovement';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!WaresOrderForm.init)  {
            WaresOrderForm.init = true;

            var w0 = LookupEditor;
            var w1 = DateEditor;
            var w2 = CounterpartiesEditor;
            var w3 = StringEditor;
            var w4 = DecimalEditor;
            var w5 = Widget;

            initFormType(WaresOrderForm, [
                'OperationTypeId', w0,
                'OrderDate', w1,
                'ShopId', w0,
                'CounterpartyId', w2,
                'EmployeeId', w0,
                'RequiredDate', w1,
                'ShipperId', w0,
                'ShippedDate', w1,
                'DocumentTypeId', w0,
                'DocumentNumber', w3,
                'DocumentDate', w1,
                'Total', w4,
                'DetailList', w5
            ]);
        }
    }
}